# MA770 Mathematical and Statistical Methods of Bioinformatics
# Problem Set 5
### Cheng, Wanli U31865818

## Genscan
### Problem 1

### (b)
Gene feeded:

GAACCATGTCCTAAATGGTTTACACTGCGCACAGCTTCCTCTCAGCCCGC
TCTGAGCTGGAAGCAGCATGTGGGACCTGGCCCTGATCTTCCTCGCAGCA
GCCTGCGTGTTCTCACTAGGGGTCACTCTGTGGGTCATTTGCAGCCATTT
TTTCACTGTGCACATCCCTGCAGCGGTTGGCCACCCTGTGAAACTGAGAG
TCCTCCATTGCATCTTCCAGCTGCTGTTGACTTGGgtgagttttgtgctt
tatgtgtcccctccagctgaccattaaggaaggcggcaggaaaaatcaca
caccggaagcttctagctgaatgaacaccggtatcatggggcctgcagtg
acagctgatcagaccttctgaaatgtgcataatccctattaggtggttct
cagcctctttgggtgtttctgaagctggccctggctaatattcacaaaat
tcaaagaatgatctgctttctcagttaacagaaagaaaaaaccagtctgg
ttctattcacctgagtgtctcccgctgactttctttttgttgttgttggt
ttgttttatttttaaaattgaggtaaaatatgcatataaaatttaccatc
tttcgcatttttaagtgtacctttcacgtgtccataaatgtgtccatatt
attcacagcactcaccactctcagaagttgtgaagatatgaatgaatttt
cctgagtctctgggcccccactgctatctacctgttagctagagccagca
agccacaaagtaaaagggtagcggatgagtagattagtgccggtgagcaa
aattacagaatgtgactttaaaaatatgttgttgcttttgagagctagaa
agctgcctagggaatatagtttggttctgcttctggtgtaatttgttaaa
atacctgtaaatggaactgtgtaggtgttgccaatacggatttcttgtac
tttttcttgtgccttgacagaaactgaatatacaaaatattttttgtaag
tttcttacataacgaaaagcaattgccagtgcctcttctcaagagacacc
ctaggtgtgtcttctatgctattcaaaataatttctatgctattcaaaag
aatgtagctattacctaaaagagaagatcccgggaactctaggcttctga
tctccagtcagcttttataagatgcagtgattgggtcctaggttttccta
cccttgatgctcagattctgtagtccaggctcccaaatacagtgactcga
tttcctcttttgcaaagtagggaaaattgctcctacttcacagggttttg
tggagactccatgaataaccctctccagaaggcttattataatgcctgac
acagagggagccctccttaaatagtagccattgttactctttgttgattc
ttttattttttatgcaaagatgtgttatgcaactgctgtgtgtcctacgc
cagtcctgggctctagggccaatgacaggtgagagaggtaaagttccaca
ggtctgcctttagggagcttgtagggtccaggcagtgggcagcggggaag
tggtctataaagaagctaacccacaatagaacaagctacagataaacagc
ttcccaaagcggagaatgcaaggaggacagcaaaacccgtcaggggagcc
gagcagggcttctctcctggggcatcttgagcccccggggccacagaaca
caagggggtcatggatttggatcacagaacataagggggtcatggatttg
ggtcacagaacacaagggggtcatggatttggatggggtcaaaattaccc
catgttcacaactaacttccaattgaaattcttctcatgaggaatgaagg
caacaaaccacagtcctgaaagccagacctcgacttttcaccaagagaaa
tcagacatatttgcattgcctattagatattgcgaaaactcatgtattct
cctcatttctttgaaactagggtagtatcaggcagcagttagatcttatt
actctatgtgctagtaaggaagaacctatattattatgacatacatttta
atattttgataagtatatttaaataacactgcttttctttataatcccat
gtatttattttatgtatttaaaaatgttctgggctgggtgtagtggctta
cacctgtaatcccagcactttgggaggctgaggtgggaggactgtttgag
tccaggagtttgagaccagcctgggtgacataatgaggcctcttctctcc
aaaaaaaaaaaaaaaaaaaaaaagttctcaaatgaggcctgcagcttttc
ccaaggccaaaatggctgtggcacagacaaggattaggaaacactggaaa
ggatacccagagggagagccctctttatttacgtatttttaactcttttt
ttttgagctggagttttactcttgttgcccaggctggagtgcaatggcat
gatctcggcttactgcaacctctgcctctagggttcaagagattcttctg
cctcagcctctcgagtagccgggaatacaggagcccaccaccacacccac
ccggctaatttttgtattttcagtagagactgggtttcaccatgttggcc
aggctggccttgaactcctgacctcaagtgacccaccagccttggcctcc
caaagtgctgggattacaggtatcagccactgtgcctggctgtattttca
accctttatgcaaactttgacatataccaaaggggaacagggtgcccact
ctccagtgtcaccatcactgactcctggcctgtctcccttccctgtgtcc
ctaaatgtctctaacttttccctcattttcatttatttaagaagccaatg
gacatcatattattttatcctaaaaattttcagtaggcattttcaaaaga
tatgaactcccccttcctgtttttaaaaaatgtacccacactatagttag
ctacacaggcttggggtggccagattggctgggacggcacaggaaggcct
ccctggaaacagatgttgcgcccaggctgaggtggagcagggcctcgtgg
tggaggtgggtgttgcaagtagagggaccatcaagggtgcccagaagcct
ggcttgtagtggctgcagtggaaggtgtcaggagagtgggttgagagggg
ctggctggtgctgtggccacctgggcctgcatgggagtggattttcctgc
aggtgtcctgagggaactgagggcttgagtagggctgtaatccaatctga
ctgcagtttccaaaaactccttttgccacctgtggagggcaggttgtgaa
ggccagactccagatggcctgtgaagaaacccatctcgacccatcatttc
ttctctctccaacagGGGATGATATTTGAGAAGCTCAGAATCTGTTCTAT
GCCCCAATTTTTCTGTTTCATGCAAGATCTGCCTCCGCTAAAGTATGACC
CCGATGTTGTGGTCACGGATTTCCGCTTTGGGACAATCCCTGTGAAGCTG
TACCAACCCAAGGCATCCACCTGCACCCTGAAGCCTGGCATCGTGTACTA
CCACGGTGGCGGGGGCGTCATGGGGAGTTTGAgtaagaaccattttctca
gacctcctaaagggtggtggcaccccttaacataacttggaagaatgggc
atcttcctgggacttaaagtatgctattattatcagggaacaccagggca
gttcatggtttgcagatcattgagggggcaaaaatatggcatatattgcc
ctcttatgtatctccttatttacataaatgtaatccttagttaaattaac
aatactgtaatataaggaaggatactgtaaggtaaagatcctgaaatgta
cccttacttgcatttatatgtgtacatatgtatgtacatataaatgtata
catgtatatttcactattttacttataatcaccacctctatttagttgga
aataaggatattttaaatgaaaagaattaaaacacagcattttgtttcac
atcaggttttgctaagacaaattctggtacagacagacaggaagatttga
gaaaaatcaatgagaggaaaaagtcactattgagacaattttactgtctt
agttattacccccagggaattaggggagaggaaacacctttatttgcttt
cagtagtgctttctaatctgtggaatgccagggtcccagtgtgggagcct
ttgagaataaaggatttaatgcaatggtggtgtggtttggtctgtatgag
aatgatagtaacagccaatatttattaagcaatattcattaatattacta
attacatgcaggcactgtgagaaccctatatgtggatgatctcattccaa
ctccaacactctacgagttagatattttcattaccccagttcacagatga
ggaaatcaagcctcaggaggttaagagacttgctaggcactatgttagct
caagctagaaaggggcaaagttgagatttgaactccagtctgaatccaga
gctcacaccctaaacctctgcgttctacagtcaaagagcttcacagatat
ttttaatggcttgtaggatggattggagggtgggcgtcttagagaaagtt
gttcaggcagtaccacgaaggagaatcagtgagaagattgaccggaagtt
tgctggagtagaggaaaacctagtcggcatcggcccaagtgctgtgtctg
taggaagaagacggtgacaatggctggcaaaggaagccttcctagtgaat
cttaaaaaccatttattttctagAAACCCACCATGGCATATGCTCTCGTT
TGTGCAAGGAGAGTGACTCCGTGGTTCTGGCAGTTGGgtgagtaaagggg
agatcccagggagccagcaaggagcaaggctctgatgtggagagatgggg
tgagaagtagaaatgggggtggggggtgggggatgggagcagatgggagc
tggaggaagcccagaggtggggatgggctgggagaagccagtgaagagag
aaaaagaaggcggctgggtgtggtggctcacgcctgtaatcccaacactt
tgggaggccacggtgggcagaatgcttgagcccaggagttcagaccagcc
tgggcaacatagtgagaccccatttttacaaaaaatacaaaaattagcca
ggtgtggtggcatatgtctgtagtcccagctacttgggaggccgaggagg
aagaagcacctgagcctgggaggttgcagtgagccgtgattgcgccacta
cactcagcctgggtgacagaacaagaccctgtcttaaaacaaacaaaaca
aaacaagaaaaagagagtgaaagaaaaataaggggaggtgaagagagatg
gagagacagagaatggggaaccccttcctctgtgcatgtgggccttgggt
ttgtttaaacagaggcgttttgtgcattttgaagctgggtaggaggtggt
cttttttaagcagttcaggtgcagagtttcactgcaggaacacttggaca
acatagctcttctttgagtaaaacaaccctgcacctccttctgctaaatg
cctgtggtaccccgcaccatcactcaaacaacccgaaatgctgccacgtt
cacttccaagtgctcccaaaagggaggtccccctggctgagacccactga
agaaagtgagaaaaacaaaaacaaaaacaaacccattgtctctcctaaca
gatctctgacagtcaccgcccagcctggagcctcaaagagggcgtggtgg
gcatggggtcccctgtcctgctgcctgatgcagcatcacacacgtggttc
ctcctggttaagtttccctgtagtttcatttaaaggtacacactttaaag
aaatatcagttctctgcatctcagtgggggtacatctgccagctttcttg
tggcacagctacctgaggtgagtagaaaacgaagagatgaagccgggtgc
agtggctcacgcttgtaatcccagcacgaggccaaggtgggtggatcact
tgaggccaggagttcaagaccagcctggccaacacggtgaaaccttgtct
ctactaaaaatacaaaaattagccaggtgtggtggcatgtgcctgtagcc
ccagctactcaggaggctgaggcaggagaatcacttgaacccagaaggca
gaggttacagtgagctgagatggcgccactgcactccagcctgggtgaca
gagtgagattccgtctgaaaaaaaaaaaaaaaaaaaaaaatggaaaagaa
aatgaagagatgagatgagagagagacctggataaacccctgactttaca
attctcagctccagaagggtttctaccatgggatcatgggatgttgccgg
acaccttgtcattctgattgtaccaaaccaagagatcgtcagtagaacat
gctagcaagtctctgccagaaagttcttttctaaatccaaggtagacagg
attttggagttaagggcattgctttccaggatgtcctattcccgggtggc
ctgaaatgggggtatggaatggaaacagctgatggaagttttgtttgttg
ctcccatagagtgtttctttgtcttttttttttttttaatttgtgaccag
catttaaaaatagtgaactttctcacaaaacctggatttctggaatctct
tagcaataccaggtaggcagggcagccatggtgttgagtgtgcatagctt
gtttggaagagtgctgtgtatctctggcctctggtggtacagtcagtcct
cagcaatgtcgacaggttcttggaactgtaacttggagtgaaaccaattt
tacctattacaaaacctaggctaattgatatagacagaatcaagctctga
tggtataattctggtcacaaaaacaccagctaatttatttaaaaattttt
taaaaatttattttactacttttatgttttataattttctttagtagaga
cgaggtcttgccatgttgcccaggctggtctcaaacttctgagctcaggc
aatccttcttcctcagcttctcagagtgctgaggttataagtgtgagcca
ctatgcctgccctaaccttctaaataaagaaaaacacctgctaatattaa
aaattgaaataaaagtgagctatacatatatttttaaaagattaaaaaaa
ataataattatttacccaatttttggtgaatcagtgagtgatggtggtca
ttctggtggtgggttaaatctaggaataaatgtttacaaggtaaaaattg
tcaggagtacctctgactgtctcccagttcaaaaccactcacaaatgtgg
cagcctcattgagagcttctgtactgcatcatttatttttgtgtagatgt
atgattattgtaaactttatgcatttttttgtttttgagatgggatctca
ccctgtcacctaggctggagtacagtggtgcgatctcagctcactgcaac
ctctgcctcctgggctcaagtgattcttgtgcctcagtctcctgagtagc
tgggaccacagacacacaccaccacgcctggctaattttttgtatttttg
gtagaggcagggtttcaccttgttgcccaggctggttttgaactcctgag
ctcaggcgattcatctgccttggcctcccaaagtgctgggattataggct
tgagccactgtgcttggctgactttataattttttttttttgagacagag
tttcactcttgttgcccaggttggagtgcaagagcgcgatctcggctcac
cacaacttctgcctcctgggttcaagggattcttctgcctcagccttcct
gagtagctgggattacaggcatgcaccaccacgcccggttaattttgtat
ttttagtagagacggggtttctccatgttggtcaggctggtcttgaactc
ctgacctcaggtgatccacctgcctcggcctcccaaagtgctgggattac
aggcgtgagccactgcacccggttggctttatgcatttttattttgcaat
aatttatattccttcatctattccttttccaactcacttattctagttca
gggtcgtgggtggctggatcctatccctgcagctcagggaacaaggcagg
aaccaaccctggacaggacacccacctattgcagggcacactcacaccta
caaccacattcacttggaccgggacaatttagacatgctcatgaatgtaa
catacacatctttgggatgtgggaggaaactggaggactgggagaaaacc
cacacagacatgaggagaatgtgcaaactccacacaaacagtagtagccc
aggctggaatcgtttttctttctttcatcaatataatgaaacaatgttga
atgaaatgacattatttgagaccctgctttttttttcttctttgagacag
ggtctcactctgtcacccaggctggagtgcagtggcatggtcacagctca
ctgcagcctcaacctcccaggcccaagtgatccacctgcctcaccctcct
gagtagctgggactacatgtatccaccaccatgcccagctaattttttat
actctttgtagaggtggggtctcaccatgttgcccaggctgggcttgaac
tcttggactcaatggatcggcccatctcagcctcccaaagtgctgaaatt
aaaggcatgagccaccggcctggccaacctgctgtatttaagtctacaag
atctcccttaacaaaatctgtaggtatggcatgattaattaaaaacccaa
gcaggtcttgaagttgctcatacaagaatcataaaacaggaagcaggagc
aggtgagtcagggaagcaacatgaggcaggttcaggaagatggggctagt
gtgggtgatatttacattgttggccaattggtggatgtctcgtgacacag
gtcccccattagggcattgagacagtaggaatggcacctgcctaggtatg
tgggctttatgtgtctcatctcatttaagcctcacacctaactttaactg
ctcaagtttcacaaagggaagtgaatggactaatttgggctaaaagatgc
tagacgctgatattttggatctgccggggcgttatcaactgtgtttcttg
attcctttttagTTACCGCAAGTTACCTAAGCATAAGTTTCCAGTGCCAG
TAAGAGACTGCTTGGTGGCCACCATCCACTTCCTGAAGTCCCTGGATGCA
TATGGAGTGGATCCAGCCCGGGTTGTGGTCTGCGGTGACAGTTTCGGAGG
GGCAATAGCCGCAGTGGTTTGTCAACAACTTGTGGACAGGCCAGATCTGC
CCCGGATCCGGGCTCAGATCCTGATCTATGCCATTCTCCAAGCCCTGGAT
TTACAAACCCCTTCGTTTCAACAGAGGAAAAACATCCCACTGCTCACCTG
GAGTTTCATCTGCTACTTTTTTTTTCAAAACCTGGATTTCAGCTCCTCCT
GGCAAGAGGTCATCATGAAAGGTGCCCATTTGCCTGCTGAAGTCTGGGAA
AAGTACAGAAAGTGGTTGGGCCCAGAAAACATCCCTGAGAGGTTTAAGGA
GAGGGGTTACCAACTGAAGCCCCATGAGCCCATGAATGAAGCTGCTTACT
TGGAAGTAAGTGTTGTCCTGGATGTGATGTGCTCGCCCCTGATTGCAGAA
GATGACATAGTGTCTCAGCTCCCGGAAACCTGCATCGTGAGCTGTGAGTA
TGATGCTCTCCGGGACAATTCACTGTTGTACAAGAAAAGGCTGGAAGACC
TGGGAGTGCCCGTGACCTGGCACCATATGGAGGATGGTTTCCATGGAGTG
CTCAGGACCATTGACATGAGCTTCTTGCACTTTCCCTGCTCCATGAGAAT
TCTGAGTGCATTAGTTCAATTTGTAAAGGGACTGTGACCATCTTTCTTCT
CTGCTGGTACTGCGGTGTGGATTCCACTGGCATCCAGCCTCCCACAGGGC
TCTCTGTTGCTGATTTAGGTGGTGCATAGTGGGGCTAGGGAGGGGGTAGA
GGTTGCTGTCACCTTTCTGGTCCAGGTTCTAGAACCACACAATGCATGCT
CCTGATGTCCAGAGGACGTGGTAGAAAAGACAGGTTTGGAGGTGGGAGTG
TGGCTGTCTCTATTCTCTGTTGGGAAAACCTGGGCTGACAATATTCAGTG
GCCATTTGTGGGAGTGAATCAGCCGGTAAGAGCTGTTCTCAGCCTCCCTA
AGGGGCAGTTCAGGCTCCCAGATTGATCCAGACTGTGTGTGACTTTCGTC
CATTTGACTTGACTTTGGAATAGCACAAGGGCATCATGTACTTCACGAGG
CTTTCCCAATGTGGCTCAGAGGCAGGAGCTCTGATGCTCTGGGCTGCTGT
GAGGTGGTGGTGGTGGTAGAGAAACTGGCTTCACCCACCTACTCTTCTGT
GAACAGTAGTGACTTTTCCCGCTGTTTCTCAGCCTCTGGGATCAGAGTCT
TCACTGTCTGGGCTGGAAACTTTAAGATAGAATGGATAGAGCTTCCACAG
TGGTTGGCATCTAGTGGTGGATGAAGACAGCCTGCAGCTGCCCGACTTGG
GGAGCTCTGGAGCTCCTGGAATCAAAGCCTGTCTTCCAACCAGAAGCCCC
AAGGCAATGTTCTAAGAATTTGAGAAGAGAAGTTGGGAGGGAAGTGGGGT
CCTGAGTTAGAGACCCATGAAGGCTGAGTCTAACCAGATAACCCTGTCCA
CAGTGCAAAGTCAAGACAGCCAAAGGAACAGAAGATGTATTTGTGAAAAC
TATTTCTTTTTTAAGACATGGAACCAACTCAAATTGGCCTCTATTAGAAA
GACAATAGATTGGCTTAGGTAGGGATGCATGCTAGGCATACATCAGGCAA
GGTTTGATCCAGGAACTCACACAGTGCCATCAGCTGTCCTGTCTTCTCTG
CTCTGCTCTTCTCTCCTCTGTGTTAATGCCACCTTCTCCTCTTCATACGG
TGGCACTGAGCAGCTTCATGCCTACCTTCCTCCAGGGTCAAGTTCATTAT
CATGGACTTGCCTCATGCTCAGCAGTCCCAGAAAAAAGCCTAATTGCAAC
TTGATGGCTTTGTTGGCTTTCTGAGCAATGTGTCCAGTTGCCACAGTGAA
GGGAATGGAATAATCTAACTCACCATTCCCAAGTCCTATGCCATCCTGAG
AGTGGGGGGTGGAGTCAATTCACCTTGGTGCTTGGACTAAGCATGAGGTG
GTGAGTGACAACGTTCCTAATTGAAGGGTAGGGTAAATGGTTGTTGGGTG
GACACCAACACTTATTCTACTACAGAAGCTAAATTGAACCCTCAGGCAGG
GTACGTGAAAGTGGCAAGAGATGTCAAGACCACTGGGCAAGTTGGCCAGT
TGTTCCTTAGGAATGAAAATTCTTTTGAAAGGAATGGCCAGGGTCCTCTG
CTGGCCCCACTTGGTCTTCTGGAGGCTCTGATCTTGGTTGGTTAGTGGTC
TTTACAGGCCAAGGTCAAGGCCATTGCACAAAAAACCCTGTGCATGCCCT
TAACTTGCTTTCAGTTGAATATTTGGGCTGAACTATGAGGCAGAGAGGAA
TCCCATTGGGTGGCTCCTTGCTGCATTCGCAGTTGACCAGCATGGGGTTT
GTTGGAGAAATAGGAACCATCCCCTGAAAACACACACTATGGTAGCCACT
CAACTGTTGAAAGGCACTGGAGTCCAATGGGTGAGGCCGCCTCTGAGACA
AGCCTCTGAGTTGAGGCTGGGAGAGGCTCCCTCCTTGGAGTGTTGCTTTT
TTTGTTTCACCCCTGCCTCTGGAGATGGGTAGAGGAACATGAGCTGACCT
TCTGGGAAGTGAGGTTGGTGAGGAGTTGCTGAGGCACTGCAGGGCCATGC
CCAGTAGAGAGGAATGTATAACATTTTAAGAGGCTGAGAGCACCCCTTGT
TGGGCGCATGCCCATGGCAGCTTCCTTCTGCCGATCATGGGAGAAATCAA
GCACTTTCACCTAATGGCTAGATGATTGATTTTGGGATGAAATTCTCCAC
TCCTCTCCTTTACCACATCACCACTATCCTTCCTGCAATACATCCACGAG
ACTCACTGAGTGGAAAAGGGATAGGAATGAATGTTCACCCAGGGCCAGCT
ACATGCTAGGCACTGTACTGGACCATTTAAATTTGCCACCTCTTATGTTC
CTCACATTAATCTTACAGAGTAGGTACAGACATACCTATGGATATTGCAG
ATTCAGTTCCAGACCACAGCAATAAAGCAAGTCACATGAATTTTTTGCTT
TCCTTAGTGCATGTAAAAGTTACATTTCCACTATATTATAGTTTATTAAG
TGTGCAATAGCATTATGTCTTTAAAAAGCATGTACATACCTTAATTTAAA
AATACCTTGTTGCTGAAAAATGCTAACAATCATCTGAGCCTTCAGTGATT
GCAGTAGCCTAGGCTACTATTTTCTATGTGGGGTTTGCACATTCTGCCCA
TGTCTGCGTGGGTTTTCTCTGAGTTCTCCAGCTTCCTCCCACATTCCAAA
GATGTGTATGTTACATTCATGGGAATGTCTAAATTGTCGTAATCTTTTTG
CTGGTTGATGGTCTTGCCTTGATGTTGATGCTGCAGGTGGTGGTTGCTGA
AGGTGGGGGAGGCTGTGGCAATTTCTTAAAATAAAATAAGACAACAGTGG
A

Output:

Sequence /tmp/10_10_19-21:26:29.fasta : 12651 bp : 46.11% C+G : Isochore 2 (43 - 51 C+G%)



Parameter matrix: HumanIso.smat



Predicted genes/exons:



Gn.Ex Type S .Begin ...End .Len Fr Ph I/Ac Do/T CodRg P.... Tscr..

----- ---- - ------ ------ ---- -- -- ---- ---- ----- ----- ------



 1.01 Init +     68    235  168  1  0   68   92    99 0.938   5.90

 1.02 Intr +   3366   3613  248  2  2   96    3   184 0.464   6.96

 1.03 Intr +   4567   4713  147  1  0   42   13   156 0.316   2.85

 1.04 Term +   9113   9887  775  2  1   92   47   437 0.999  32.86

 1.05 PlyA +  12221  12226    6                               1.05



 2.02 PlyA -  12297  12292    6                               1.05

 2.01 Term -  12580  12392  189  1  0   65   49   114 0.600   2.65



Suboptimal exons with probability > 0.010



Exnum Type S .Begin ...End .Len Fr Ph B/Ac Do/T CodRg P.... Tscr..

----- ---- - ------ ------ ---- -- -- ---- ---- ----- ----- ------



S.001 Intr +   3026   3055   30  1  0    7   90    35 0.014  -6.07

S.002 Intr +   3366   3555  190  2  1   96   -5   190 0.020   9.66

S.003 Intr +   3366   3582  217  2  1   96   43   191 0.283  13.48

S.004 Intr +   3366   3616  251  2  2   96  -17   189 0.162   5.44

S.005 Intr +   3427   3613  187  2  1   37    3   210 0.017   6.15

S.006 Intr +   3444   3613  170  2  2    4    3   203 0.017   2.14

S.007 Intr +   4531   4688  158  1  2   71   -8   133 0.010   1.65

S.008 Intr +   4531   4713  183  1  0   71   13   126 0.217   2.30

S.009 Term +   4531   4717  187  1  1   71   48   132 0.017   4.46

S.010 Intr +   4567   4688  122  1  2   42   -8   163 0.016   2.24

S.011 Intr +   4567   4699  133  1  1   42   -3   159 0.011   2.85

S.012 Term +   4567   4717  151  1  1   42   48   162 0.025   4.98

S.013 Intr +   4593   4713  121  1  1  -46   13   157 0.062  -5.65

S.014 Intr +   4774   4837   64  2  1   56   92     8 0.186  -4.22

S.015 Intr +   5836   5917   82  2  1   64  105   -38 0.062  -5.80

S.016 Init +   6678   6736   59  2  2  106   37    11 0.028  -3.25

S.017 Init +   8840   8850   11  1  2   82   15     6 0.014  -7.25

S.018 Term -  12596  12392  205  1  1   41   49   131 0.243   1.34

S.019 Term -  12596  12529   68  0  2   41   42    85 0.015  -2.70





Predicted peptide sequence(s):





>/tmp/10_10_19-21:26:29.fasta|GENSCAN_predicted_peptide_1|445_aa

MWDLALIFLAAACVFSLGVTLWVICSHFFTVHIPAAVGHPVKLRVLHCIFQLLLTWGMIF

EKLRICSMPQFFCFMQDLPPLKYDPDVVVTDFRFGTIPVKLYQPKASTCTLKPGIVYYHG

GGGVMGSLSKNHFLRPPKGMDWRVGVLEKVVQAVPRRRISEKIDRKFAGVEENLVGIGPS

AVSVGRRRYRKLPKHKFPVPVRDCLVATIHFLKSLDAYGVDPARVVVCGDSFGGAIAAVV

CQQLVDRPDLPRIRAQILIYAILQALDLQTPSFQQRKNIPLLTWSFICYFFFQNLDFSSS

WQEVIMKGAHLPAEVWEKYRKWLGPENIPERFKERGYQLKPHEPMNEAAYLEVSVVLDVM

CSPLIAEDDIVSQLPETCIVSCEYDALRDNSLLYKKRLEDLGVPVTWHHMEDGFHGVLRT

IDMSFLHFPCSMRILSALVQFVKGL



>/tmp/10_10_19-21:26:29.fasta|GENSCAN_predicted_peptide_2|62_aa

HQHQGKTINQQKDYDNLDIPMNVTYTSLECGRKLENSEKTHADMGRMCKPHIENSSLGYC

NH

## HMM
### Problem 2

In [271]:
import numpy as np

def O(s):
    ### This function create the string O_t of observables.
    vec = np.zeros(len(s)+2)
    vec[len(s)+1]=5
    for n in range(0,len(s)):
        if s[n] == 'A':
            vec[n+1] = 1
        elif s[n] == 'C':
            vec[n+1] = 2
        elif s[n] == 'G':
            vec[n+1] = 3
        elif s[n] == 'T':
            vec[n+1] = 4
    return vec

P1 = np.array([
    [0,0.3,0.7,0,0,0],
    [0,0.7,0.3,0,0,0],
    [0,0,0,0.4,0.6,0],
    [0,0,0,0.5,0.5,0],
    [0.5,0,0,0,0,0.5],
    [0.7,0,0,0,0,0.3]
])
### This is our initial guess of the transition probability.
###'Begin' and 'End' states are added.

Q1 = np.array([
    [1,0,0,0,0],
    [0,0.7,0.1,0.1,0.1],
    [0,0.2,0.4,0.2,0.2],
    [0,0.1,0.1,0.7,0.1],
    [0,0.2,0.4,0.2,0.2],
    [0,0.1,0.1,0.1,0.7]
])
### This is our initial guess of the emission probability. 
###'Begin' and 'End' states are added.

def F(P,Q,O,i,t):
    ### This function calculate the forward probability.
    (n,m) = np.shape(Q1)
    if t > 0:
        ### The general case.
        return Q[i,np.int(O[t])]* sum(P[j,i]* F(P,Q,O,j,t-1) for j in range(0,n))
    else:
        ### The base case.
        return Q[i,np.int(O[0])]
        
def B(P,Q,O,i,t):
    ### This function calculate the backward probability.
    (n,m) = np.shape(Q1)
    l=len(O)-1
    if  t < l-1:
        ### The general case.
        return sum(P[i,j]* Q[j,np.int(O[t+1])]* B(P,Q,O,j,t+1) for j in range(0,n))
    elif t == l-1:
        ### The base case.
        return P[i,l-1]
    else:
        return 0

def FResults(P,Q,O):
    ### This function collect all the forward probabilities.
    (n,m) = np.shape(Q1)
    l=len(O)-1
    M=np.zeros((n,l))
    for i in range(0,n):
        for t in range(0,l):
            M[i,t]=F(P,Q,O,i,t)
    return M

def BResults(P,Q,O):
    ### This function collect all the forward probabilities.
    (n,m) = np.shape(Q1)
    l=len(O)-1
    M=np.zeros((n-1,l))
    for i in range(0,n-1):
        for t in range(0,l):
            M[i,t]=B(P,Q,O,i,t)
    return M

### (a)
The transition matrix is given by:
$$
P=
\begin{bmatrix}
0& 0.3& 0.7& 0& 0& 0\\
0& 0.7& 0.3& 0& 0& 0\\
0& 0& 0& 0.4& 0.6& 0\\
0& 0& 0& 0.5& 0.5& 0\\
0.5& 0& 0& 0& 0& 0.5\\
0.7& 0& 0& 0& 0& 0.3
\end{bmatrix}
$$
### (b)
The emmision matrix is given by:
$$
Q=
\begin{bmatrix}
1& 0& 0& 0& 0\\
0& 0.7& 0.1& 0.1& 0.1\\
0& 0.2& 0.4& 0.2& 0.2\\
0& 0.1& 0.1& 0.7& 0.1\\
0& 0.2& 0.4& 0.2& 0.2\\
0& 0.1& 0.1& 0.1& 0.7\\
\end{bmatrix}
$$

### (c)
Randomly generated sequences: 
1. "GACG"
2. "ACAT".

In [272]:
p1=0.3*0.1*(0.7*0.7*(0.7*0.1*(0.7*0.1+0.3*0.2)+0.3*0.2*(0.4*0.7+0.6*0.2))+0.3*0.2*(0.4*0.1*(0.5*0.7+0.5*0.2)+0.6*0.2*(0.5*0+0.5*0.1)))+0.7*0.2*(0.4*0.1*(0.5*0.1*(0.5*0.2+0.5*0.7)+0.5*0.2*(0.5*0+0.5*0.1))+0.6*0.2*(0.5*0+0.5*0.7*(0.3*0.1+0.7*0)))
p2=0.3*0.7*(0.7*0.1*(0.7*0.7*(0.7*0.1+0.3*0.4)+0.3*0.2*(0.4*0.1+0.6*0.4))+0.3*0.2*(0.4*0.1*(0.5*0.1+0.5*0.5)+0.6*0.2*(0.5*0.1)))+0.7*0.2*(0.4*0.1*(0.5*0.1*(0.5*0.1+0.5*0.4)+0.5*0.2*(0.5*0.1))+0.6*0.2*(0.5*0.1*(0.3*0.1)))

In [273]:
p1

0.0008601699999999999

In [274]:
p2

0.00196553

The probability of the sequences are
\begin{align}
Prob(GACG)&=0.0008601699999999999\\
Prob(ACAT)&=0.00196553
\end{align}

### (d)

In [275]:
s1="GACG"
s2="ACAT"
O1=O(s1)
O2=O(s2)

The forward and backward probabilities of "GACG" is given by:

In [277]:
FResults(P1,Q1,O1)

array([[1.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00],
       [0.0000e+00, 3.0000e-02, 1.4700e-02, 1.0290e-03, 7.2030e-05],
       [0.0000e+00, 1.4000e-01, 1.8000e-03, 1.7640e-03, 6.1740e-05],
       [0.0000e+00, 0.0000e+00, 5.6000e-03, 3.5200e-04, 6.1712e-04],
       [0.0000e+00, 0.0000e+00, 1.6800e-02, 1.5520e-03, 2.4688e-04],
       [0.0000e+00, 0.0000e+00, 0.0000e+00, 8.4000e-04, 1.0280e-04]])

In [278]:
BResults(P1,Q1,O1)

array([[3.45604e-04, 5.03720e-03, 4.02800e-02, 8.40000e-02, 0.00000e+00],
       [7.13076e-04, 9.88680e-03, 1.93200e-02, 3.60000e-02, 0.00000e+00],
       [1.22500e-04, 3.50000e-04, 7.00000e-03, 1.40000e-01, 6.00000e-01],
       [1.53125e-04, 4.37500e-04, 8.75000e-03, 1.75000e-01, 5.00000e-01],
       [0.00000e+00, 0.00000e+00, 0.00000e+00, 0.00000e+00, 0.00000e+00]])

The forward and backward probabilities of "ACAT" is given by:

In [279]:
FResults(P1,Q1,O2)

array([[1.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00],
       [0.0000e+00, 2.1000e-01, 1.4700e-02, 7.2030e-03, 5.0421e-04],
       [0.0000e+00, 1.4000e-01, 2.5200e-02, 8.8200e-04, 4.3218e-04],
       [0.0000e+00, 0.0000e+00, 5.6000e-03, 1.2880e-03, 9.9680e-05],
       [0.0000e+00, 0.0000e+00, 3.3600e-02, 3.5840e-03, 2.3464e-04],
       [0.0000e+00, 0.0000e+00, 0.0000e+00, 1.6800e-03, 1.6072e-03]])

In [280]:
BResults(P1,Q1,O2)

array([[3.09148e-04, 8.45200e-04, 1.03600e-02, 8.40000e-02, 0.00000e+00],
       [7.08012e-04, 1.43880e-03, 1.88400e-02, 3.60000e-02, 0.00000e+00],
       [2.50000e-06, 5.00000e-05, 1.00000e-03, 2.00000e-02, 6.00000e-01],
       [3.12500e-06, 6.25000e-05, 1.25000e-03, 2.50000e-02, 5.00000e-01],
       [0.00000e+00, 0.00000e+00, 0.00000e+00, 0.00000e+00, 0.00000e+00]])

## Matrix Limit
### Problem 3

In [136]:
M1 = np.array([
    [.7,.1,.1,.1],
    [.1,.7,.1,.1],
    [.1,.1,.7,.1],
    [.1,.1,.1,.7],
])

The transition matrix is given by
$$
M = 
\begin{bmatrix}
    .7&.1&.1&.1,\\
    .1&.7&.1&.1,\\
    .1&.1&.7&.1,\\
    .1&.1&.1&.7,
\end{bmatrix}
$$

### (a)
$M^2$ is given by:

In [137]:
np.linalg.matrix_power(M1,2)

array([[0.52, 0.16, 0.16, 0.16],
       [0.16, 0.52, 0.16, 0.16],
       [0.16, 0.16, 0.52, 0.16],
       [0.16, 0.16, 0.16, 0.52]])

$$
M^2=
\begin{bmatrix}
0.52& 0.16& 0.16& 0.16\\
0.16& 0.52& 0.16& 0.16\\
0.16& 0.16& 0.52& 0.16\\
0.16& 0.16& 0.16& 0.52
\end{bmatrix}
$$

$M^{16}$ is given by:

In [138]:
np.linalg.matrix_power(M1,16)

array([[0.25021158, 0.24992947, 0.24992947, 0.24992947],
       [0.24992947, 0.25021158, 0.24992947, 0.24992947],
       [0.24992947, 0.24992947, 0.25021158, 0.24992947],
       [0.24992947, 0.24992947, 0.24992947, 0.25021158]])

$$
M^{16}=
\begin{bmatrix}
0.25021158& 0.24992947& 0.24992947& 0.24992947\\
0.24992947& 0.25021158& 0.24992947& 0.24992947\\
0.24992947& 0.24992947& 0.25021158& 0.24992947\\
0.24992947& 0.24992947& 0.24992947& 0.25021158
\end{bmatrix}
$$

And $M^{256}$ is given by:

In [139]:
np.linalg.matrix_power(M1,256)

array([[0.25, 0.25, 0.25, 0.25],
       [0.25, 0.25, 0.25, 0.25],
       [0.25, 0.25, 0.25, 0.25],
       [0.25, 0.25, 0.25, 0.25]])

$$
M^{256}=
\begin{bmatrix}
0.25& 0.25& 0.25& 0.25\\
0.25& 0.25& 0.25& 0.25\\
0.25& 0.25& 0.25& 0.25\\
0.25& 0.25& 0.25& 0.25
\end{bmatrix}
$$

### (b)
Diagonalize M

In [140]:
np.linalg.eig(M1)

(array([0.6, 1. , 0.6, 0.6]),
 array([[-0.8660254 ,  0.5       , -0.35510142, -0.30663851],
        [ 0.28867513,  0.5       ,  0.83717124, -0.27020982],
        [ 0.28867513,  0.5       , -0.40963449, -0.28888965],
        [ 0.28867513,  0.5       , -0.07243534,  0.86573798]]))

Therefore, by definition we have:
\begin{align}
D&=
\begin{bmatrix}
1&0&0&0\\
0&0.6&0&0\\
0&0&0.6&0\\
0&0&0&0.6
\end{bmatrix}
\\
P&=
\begin{bmatrix}
1/2&-0.8660254&-0.35510142&-0.30663851\\
1/2&0.28867513&0.83717124&-0.27020982\\
1/2&0.28867513&-0.40963449&-0.28888965\\
1/2&0.28867513&-0.07243534&0.86573798
\end{bmatrix}
\end{align}

Moreover,
\begin{align}
\lim_{n\to\infty}M^{n}\mathbf{x}&=\lim_{n\to\infty}(PDP^{T})^n\mathbf{x}\\
&=P(\lim_{n\to\infty}D^n)P^{T}\mathbf{x}\\
&=P
\begin{bmatrix}
1&&&\\
&0&&\\
&&0&\\
&&&0
\end{bmatrix}
P^{T}\mathbf{x}\\
&=\begin{bmatrix}
1/2&0&0&0\\
1/2&0&0&0\\
1/2&0&0&0\\
1/2&0&0&0
\end{bmatrix}
P^{T}\mathbf{x}\\
&=\begin{bmatrix}
1/4&1/4&1/4&1/4\\
1/4&1/4&1/4&1/4\\
1/4&1/4&1/4&1/4\\
1/4&1/4&1/4&1/4
\end{bmatrix}
\mathbf{x}\\
&=\begin{bmatrix}
(x^1+x^2+x^3+x^4)/4\\
(x^1+x^2+x^3+x^4)/4\\
(x^1+x^2+x^3+x^4)/4\\
(x^1+x^2+x^3+x^4)/4
\end{bmatrix}
\mathbf{x}\\
&=\begin{bmatrix}
1/4\\
1/4\\
1/4\\
1/4
\end{bmatrix}
\end{align}

## Jukes-Cantor model
### Problem 4

In [141]:
a = 1/6
Ma = np.array([
    [1-3*a,a,a,a],
    [a,1-3*a,a,a],
    [a,a,1-3*a,a],
    [a,a,a,1-3*a],
])

Given the evolution matrix:
$$
M=
\begin{bmatrix}
1-3\alpha&\alpha&\alpha&\alpha \\
\alpha&1-3\alpha&\alpha&\alpha \\
\alpha&\alpha&1-3\alpha&\alpha \\
\alpha&\alpha&\alpha&1-3\alpha \\
\end{bmatrix}
$$

### (a)

In [142]:
np.linalg.eig(Ma)

(array([0.33333333, 1.        , 0.33333333, 0.33333333]),
 array([[-0.8660254 ,  0.5       , -0.11365536,  0.19200516],
        [ 0.28867513,  0.5       ,  0.84731974, -0.22014476],
        [ 0.28867513,  0.5       , -0.36683219, -0.66204941],
        [ 0.28867513,  0.5       , -0.36683219,  0.69018902]]))

The eigen values are given by
\begin{align}
D=
\begin{bmatrix}
\lambda_1&&&\\
&\lambda_2&&\\
&&\lambda_3&\\
&&&\lambda_4\\
\end{bmatrix}=
\begin{bmatrix}
1&&&\\
&1/3&&\\
&&1/3&\\
&&&1/3\\
\end{bmatrix},
\end{align}
and the corresponding eigenvalues are:
\begin{align}
P=
\begin{bmatrix}
\mathbf{v}_1&\mathbf{v}_2&\mathbf{v}_3&\mathbf{v}_4
\end{bmatrix}=
\begin{bmatrix}
1/2&-0.8660254&-0.11365536&0.19200516\\
1/2&0.28867513&0.84731974&-0.22014476\\
1/2&0.28867513&-0.36683219&-0.66204941\\
1/2&0.28867513&-0.36683219&0.69018902\\
\end{bmatrix},
\end{align}

### (b)

#### Cliam:
For arbitrary $\alpha$ s.t., $0\le\alpha\le1$, there exists one eigenvalue of $M$ s.t. $\lambda_i=1$.

#### Proof:
For some $\lambda_i=1$, we have:
\begin{align}
|M-\lambda_{i}I|&=
\begin{vmatrix}
-3\alpha&\alpha&\alpha&\alpha\\
\alpha&-3\alpha&\alpha&\alpha\\
\alpha&\alpha&-3\alpha&\alpha\\
\alpha&\alpha&\alpha&-3\alpha\\
\end{vmatrix}
\\
&=
-3\alpha\begin{vmatrix}
-3\alpha&\alpha&\alpha\\
\alpha&-3\alpha&\alpha\\
\alpha&\alpha&-3\alpha\\
\end{vmatrix}-
\alpha
\begin{vmatrix}
\alpha&\alpha&\alpha\\
\alpha&-3\alpha&\alpha\\
\alpha&\alpha&-3\alpha\\
\end{vmatrix}+
\alpha
\begin{vmatrix}
\alpha&-3\alpha&\alpha\\
\alpha&\alpha&\alpha\\
\alpha&\alpha&-3\alpha\\
\end{vmatrix}-
\alpha
\begin{vmatrix}
\alpha&-3\alpha&\alpha\\
\alpha&\alpha&-3\alpha\\
\alpha&\alpha&\alpha\\
\end{vmatrix}
\\
&=
-3\alpha[-3\alpha(9\alpha^2-\alpha^2)+\alpha(3\alpha^2+\alpha^2)+\alpha(\alpha^2+3\alpha^2)]+\\
&\quad
-\alpha[\alpha(9\alpha^2-\alpha^2)+\alpha(3\alpha^2+\alpha^2)+\alpha(\alpha^2+3\alpha^2)]+\\
&\quad
\alpha[\alpha(-3\alpha^2-\alpha^2)+3\alpha(-3\alpha^2-\alpha^2)+\alpha(\alpha^2-\alpha^2)]+\\
&\quad
-\alpha[\alpha(\alpha^2+3\alpha^2)+3\alpha(\alpha^2+3\alpha^2)+\alpha(\alpha^2-\alpha^2)]
\\
&=
48\alpha^3-16\alpha^3-16\alpha^3-16\alpha^3\\
&=0,
\end{align}
which implies $\lambda_i=1$ is indead an eigenvalue of $M$ by definition.

Its corresponding eigenvector satisfies the following expression:
$$
M\mathbf{v}_i=\mathbf{v}_i
$$
hence we have
\begin{align}
\begin{bmatrix}
(1-3\alpha) v_i^1+\alpha v_i^2+\alpha v_i^3+\alpha v_i^4\\
\alpha v_i^1+(1-3\alpha) v_i^2+\alpha v_i^3+\alpha v_i^4\\
\alpha v_i^1+\alpha v_i^2+(1-3\alpha) v_i^3+\alpha\\
\alpha v_i^1+\alpha v_i^2 + \alpha v_i^1 + (1-3\alpha)v_i^4\\
\end{bmatrix}=
\begin{bmatrix}
v_i^1\\
v_i^2\\
v_i^3\\
v_i^4\\
\end{bmatrix}
\end{align}
which implies 
\begin{cases}
-3\alpha v_i^1+\alpha v_i^2+\alpha v_i^3+\alpha v_i^4=0\\
\alpha v_i^1-3\alpha v_i^2+\alpha v_i^3+\alpha v_i^4=0\\
\alpha v_i^1+\alpha v_i^2-3\alpha v_i^3+\alpha v_i^4=0\\
\alpha v_i^1+\alpha v_i^2+\alpha v_i^3-3\alpha v_i^4=0\\
\end{cases}
which can be solved as
$$
\mathbf{v}_i=k
\begin{bmatrix}
1/2\\
1/2\\
1/2\\
1/2\\
\end{bmatrix}
$$
for arbitrary $k$.

## Kimura Model
### Problem 5
Let $\alpha, \beta>0$ and $\alpha+2\beta<1$, the transition matrix is given by:
$$
M=
\begin{bmatrix}
1-\alpha-2\beta&\alpha&\beta&\beta \\
\alpha&1-\alpha-2\beta&\beta&\beta \\
\beta&\beta&1-\alpha-2\beta&\alpha \\
\beta&\beta&\alpha&1-\alpha-2\beta \\
\end{bmatrix}
$$

### （a）
To verified that

\begin{align}
M&=
\frac{1}{4}
\begin{bmatrix}
1&1&1&1\\
1&1&1&1\\
1&1&1&1\\
1&1&1&1
\end{bmatrix}
+
\frac{1-4\beta}{4}
\begin{bmatrix}
1&1&-1&-1\\
1&1&-1&-1\\
-1&-1&1&1\\
-1&-1&1&1
\end{bmatrix}
+
\frac{1-2(\alpha+\beta)}{2}
\begin{bmatrix}
1&-1&0&0\\
-1&1&0&0\\
0&0&1&-1\\
0&0&-1&1
\end{bmatrix}
,
\\
\text{we have}
\\
RHS
&=
\frac{1}{4}
\begin{bmatrix}
2-4\beta&2-4\beta&4\beta&4\beta\\
2-4\beta&2-4\beta&4\beta&4\beta\\
4\beta&4\beta&2-4\beta&2-4\beta\\
4\beta&4\beta&2-4\beta&2-4\beta\\
\end{bmatrix}
+
\frac{1}{4}
\begin{bmatrix}
2-4\alpha-4\beta&4\alpha+4\beta-2&0&0\\
4\alpha+4\beta-2&2-4\alpha-4\beta&0&0\\
0&0&2-4\alpha-4\beta&4\alpha+4\beta-2\\
0&0&4\alpha+4\beta-2&2-4\alpha-4\beta
\end{bmatrix}
\\
&=
\frac{1}{4}
\begin{bmatrix}
4-4\alpha-8\beta&4\alpha&4\beta&4\beta\\
4\alpha&4-4\alpha-8\beta&4\beta&4\beta\\
4\beta&4\beta&4-4\alpha-8\beta&4\alpha\\
4\beta&4\beta&4\alpha&4-4\alpha-8\beta
\end{bmatrix}
\\
&=
\begin{bmatrix}
1-\alpha-2\beta&\alpha&\beta&\beta\\
\alpha&1-\alpha-2\beta&\beta&\beta\\
\beta&\beta&1-\alpha-2\beta&\alpha\\
\beta&\beta&\alpha&1-\alpha-2\beta
\end{bmatrix}
\\
&=LHS
.
\end{align}

### （b）
For
\begin{align}
&\mathbf{v}_{1}=
\frac{1}{2}
\begin{bmatrix}
1\\
1\\
1\\
1
\end{bmatrix}
&&\mathbf{v}_{2}=
\frac{1}{2}
\begin{bmatrix}
1\\
1\\
-1\\
-1
\end{bmatrix}
&&\mathbf{v}_{3}=
\frac{1}{2}
\begin{bmatrix}
1\\
-1\\
1\\
-1
\end{bmatrix}
&&\mathbf{v}_{4}=
\frac{1}{2}
\begin{bmatrix}
1\\
-1\\
-1\\
1
\end{bmatrix},
\end{align}

we immediately have

\begin{align}
\mathbf{v}_{1}\mathbf{v}_{1}^{T}&=
\frac{1}{4}
\begin{bmatrix}
1&1&1&1\\
1&1&1&1\\
1&1&1&1\\
1&1&1&1\\
\end{bmatrix}
\\
\mathbf{v}_{2}\mathbf{v}_{2}^{T}&=
\frac{1}{4}
\begin{bmatrix}
1&1&-1&-1\\
1&1&-1&-1\\
-1&-1&1&1\\
-1&-1&1&1
\end{bmatrix}
\\
\mathbf{v}_{3}\mathbf{v}_{3}^{T}+\mathbf{v}_{4}\mathbf{v}_{4}^{T}&=
\frac{1}{4}
\begin{bmatrix}
1&-1&1&-1\\
-1&1&-1&1\\
1&-1&1&-1\\
-1&1&-1&1
\end{bmatrix}
+
\frac{1}{4}
\begin{bmatrix}
1&-1&-1&1\\
-1&1&1&-1\\
-1&1&1&-1\\
1&-1&-1&1
\end{bmatrix}
\\
&=
\frac{1}{4}
\begin{bmatrix}
2&-2&0&0\\
-2&2&0&0\\
0&0&2&-2\\
0&0&-2&2
\end{bmatrix}
\\
&=
\frac{1}{2}
\begin{bmatrix}
1&-1&0&0\\
-1&1&0&0\\
0&0&1&-1\\
0&0&-1&1
\end{bmatrix}
\end{align},

which verifies the identity:
$$
M=
1\cdot\mathbf{v}_{1}\mathbf{v}_{1}^{T}
+(1-4\beta)\cdot\mathbf{v}_{2}\mathbf{v}_{2}^{T}
+(1-2(\alpha+\beta))(\mathbf{v}_{3}\mathbf{v}_{3}^{T}+\mathbf{v}_{4}\mathbf{v}_{4}^{T})
$$

### (d)
Let $M$ denote an arbitrary matrix with a complete orthonormal set of eigenvectors $\lambda_i$ and eigenvectors $\mathbf{v_i}$.

Show that
$$
M=\sum_{i=1}^{n}\lambda_i\mathbf{v}_{i}\mathbf{v}_{i}^{T}
$$

#### Proof:
In the given context, we can write $M$ in the form of $PDP^{T}$ s.t.
\begin{align}
P&=
\begin{bmatrix}
\mathbf{v}_{1}&\dots&\mathbf{v}_{n}
\end{bmatrix}
\\
D&=
\begin{bmatrix}
\lambda_1&&\\
&\ddots&\\
&&\lambda_n
\end{bmatrix}
\end{align}
without loss of generality.

Therefore, we have the following expression:

\begin{align}
M&=
PDP^{T}\\
&=
\begin{bmatrix}
\lambda_1\mathbf{v}_{1}&\dots&\lambda_n\mathbf{v}_{n}
\end{bmatrix}
\begin{bmatrix}
\mathbf{v}_{1}\\
\vdots\\
\mathbf{v}_{n}
\end{bmatrix}
\\
&=
\begin{bmatrix}
\sum_{i=1}^{n}\lambda_{i}v_{i}^{1}v_{i}^{1}&\dots&\sum_{i=1}^{n}\lambda_i v_{i}^{1}v_{i}^{n}\\
\vdots&\ddots&\vdots\\
\sum_{i=1}^{n}\lambda_{i}v_{i}^{n}v_{i}^{1}&\dots&\sum_{i=1}^{n}\lambda_i v_{i}^{n}v_{i}^{n}
\end{bmatrix}
\\
&=
\sum_{i=1}^{n}\lambda_{i}
\begin{bmatrix}
v_{i}^{1}v_{i}^{1}&\dots&v_{i}^{1}v_{i}^{n}\\
\vdots&\ddots&\vdots\\
v_{i}^{n}v_{i}^{1}&\dots&v_{i}^{n}v_{i}^{n}
\end{bmatrix}
\\
&=\sum_{i=1}^{n}\lambda_i\mathbf{v}_{i}\mathbf{v}_{i}^{T},
\end{align}
where $v_{i}^{j}$ denote the $j^{th}$ component of the vector $\mathbf{v}_{i}$.

Hence our claim is verified.

What is the interpretation of $\mathbf{v}_{i}\mathbf{v}_{i}^{T}$ ?

#### Answer:
It could be interpretated as an orthogonal basis of the matrix $M$. In another word, the transition matrix could be interprated as a linear combination of several basic modes of evolution behaviors which are orthogonal with each other. And each of the basic modes is represented by $\mathbf{v}_{i}\mathbf{v}_{i}^{T}$. $\lambda_{i}$ could be interprated as the significance of some mode in a evolution process.

## Markov Chain: Reversibility
### Problem 6
Let $\{\psi(t)\}$ denote a Markov chain with possible underlying states $\{\varphi_1,\dots,\varphi_m\}$.

The initial distribution is selected to agree with the stationary distribution $\pi=(\pi_{k})$ which is given by,
$$
\pi_{k}=Prob(\psi(0)=\varphi_k)=\lim_{t\to\infty}Prob(\psi(t)=\varphi_k).
$$

The transition matrix is given by $P=(p_{ij})$ s.t.,
$$
p_{ij}=Prob(\psi(t+1)=\varphi_j|\psi(t)=\varphi_i)
$$

In addition we assume this chain to be irreducible and periodic.

Define the reverse sequence of $\{\psi(t)\}$ to be $\{\psi^{*}(t)\}$ s.t.,
$$
\psi^{*}(t)=\psi(T-t).
$$
### (a)
#### Theorem 1:
The prebabilistic structure of $\psi^{*}(t)$ is also regular Markov chain with transition matrix $P^{*}=(p^{*}_{ij})$ given by
$$
p^{*}_{ij}=p_{ji}\frac{\pi_{j}}{\pi_i}.
$$
Furthermore, the stationary distribution of reversed chain $\psi^{*}(t)$ is also $\pi$.

#### Proof:
By definition, we have
\begin{align}
p^{*}_{ij}&=Prob(\psi(t)=\varphi_j|\psi(t+1)=\varphi_i)\\
&=Prob(\psi(t+1)=\varphi_i|\psi(t)=\varphi_j)\frac{Prob(\psi(t)=\varphi_j)}{Prob(\psi(t+1)=\varphi_i)}\quad\text{(By Bayes' Theorem)}\\
&=p_{ji}\frac{\pi_{j}(t)}{\pi_{i}(t+1)}
\end{align}
By definition, we have:
$$
\lim_{t\to\infty}p_{ij}=p_{ij},
$$
and
$$
\lim_{t\to\infty}p^{*}_{ij}=p^{*}_{ij}.
$$
Therefore,
\begin{align}
p^{*}_{ij}
&=\lim_{t\to\infty}p^{*}_{ij}=p^{*}_{ij}\\
&=\lim_{t\to\infty}(p_{ji}\frac{\pi_{j}(t)}{\pi_{i}(t+1)})\\
&=p_{ji}\lim_{t\to\infty}\frac{\pi_{j}(t)}{\pi_{i}(t+1)}\\
&=p_{ji}\frac{\pi_{j}}{\pi_{i}}
\end{align}

Furthermore, the stationary distribution of the reversed chain is given by:
\begin{align}
\pi^{*}_j&=(\lim_{t\to\infty}\pi^{*}(0){P^{*}}^{t})_{j}\\
&=\lim_{t\to\infty}\sum_{i=1}^{n}\pi_{i}^{*}(0){P^{*}}^{t}_{ij}
\end{align}

Since
\begin{align}
p^{*}_{ij}p^{*}_{jk}&=p_{ji}\frac{\pi_{j}}{\pi_{i}}p_{kj}\frac{\pi_{k}}{\pi_{j}}\\
&=p_{kj}p_{ji}\frac{\pi_{k}}{\pi_{i}}
\end{align}
we can deduce that
\begin{align}
{P^{*}}^{t}_{ij}&={{P}^{t}}_{ji}\frac{\pi_{j}}{\pi_{i}}
\end{align}
by definition of matrix multiplication.

In addition, by construction of the reversed chain, we have:
\begin{align}
\pi_{i}^{*}(0)=\pi_{i}
\end{align}

Therefore, we have:
\begin{align}
\pi^{*}_j&=\lim_{t\to\infty}\sum_{i=1}^{n}\pi_{i}{{P}^{t}}_{ji}\frac{\pi_{j}}{\pi_{i}}\\
&=\lim_{t\to\infty}\sum_{i=1}^{n}\pi_{j}{{P}^{t}}_{ji}\\
&=\lim_{t\to\infty}[\pi{P}^{t}]_{j}\\
&=\pi_{j}
\end{align},
which verified our claim s,t, $\pi^*=\pi$.

### (b)
#### Proposition: 
If $\{\psi(t)\}$ is irreducible, then $\{\psi^{*}(t)\}$ is also irreducible.

#### Proof:
By hypothesis, we have:
\begin{align}
\forall i,j\ \exists k\ s.t.\ Prob(\psi_{t+k}=\varphi_j|\psi_t=\varphi_i)>0
\end{align}
Therefore, we must have $\pi_j>0$ for all $j$, otherwise, we cannot reach $\varphi_j$ in finite steps.

By construction, we also have:
\begin{align}
Prob(\psi_{t+k}=\varphi_j|\psi_t=\varphi_i)\pi_i=Prob(\psi_{t+k}=\varphi_j,\psi_t=\varphi_i)
\end{align}
i.e., $\{\psi_{t}\}$ is irreducible iff $Prob(\psi_{t+k}=\varphi_j,\psi_t=\varphi_i)>0$.

Then consider the reversed chain $\{\psi^{*}_{t}\}$ (without loss of generality, we assume the reversed chain start at $T=t+k$):

\begin{align}
Prob(\psi^{*}_{k}=\varphi_j,\psi^{*}_{0}=\varphi_i)&=Prob(\psi_{t}=\varphi_j,\psi_{t+k}=\varphi_i)\\
&=Prob(\psi_{t+k}=\varphi_j,\psi_{t}=\varphi_i)\\
&>0
\end{align}

Then, we have
\begin{align}
Prob(\psi^{*}_{k}=\varphi_j|\psi^{*}_0=\varphi_i)\pi^*_i=Prob(\psi^{*}_{k}=\varphi_j,\psi^{*}_{0}=\varphi_i)>0.
\end{align}
Since $\pi^*_i=\pi_i>0$, we can deduce that
$$
Prob(\psi^{*}_{k}=\varphi_j|\psi^{*}_0=\varphi_i)>0,
$$
which implies the reversed chain is irreducible.